File: composer/src/main/java/io/piano/android/composer/DependenciesProvider.java
package io.piano.android.composer;

import android.content.Context;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.concurrent.TimeUnit;

import io.piano.android.composer.model.Event;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@RestrictTo(RestrictTo.Scope.LIBRARY)
class DependenciesProvider {
    private static DependenciesProvider instance;

    private final Composer composer;

    private DependenciesProvider(@NonNull Context context, @NonNull String aid, @Nullable String endpoint) {
        Context appContext = context.getApplicationContext();

        PrefsStorage prefsStorage = new PrefsStorage(appContext);

        String userAgent = buildUserAgent(appContext);

        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .addInterceptor(new UserAgentInterceptor(userAgent))
                .addInterceptor(loggingInterceptor)
                .build();

        Gson gson = new GsonBuilder()
                .setLenient()
                .registerTypeAdapter(Event.class, new EventDeserializer())
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(Constants.BASE_URL_SANDBOX)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
        Api api = retrofit.create(Api.class);
        ExperienceIdsProvider experienceIdsProvider = new ExperienceIdsProvider(prefsStorage);
        HttpHelper httpHelper = new HttpHelper(experienceIdsProvider, prefsStorage, gson, userAgent);
        composer = new Composer(api, httpHelper, aid, endpoint);
    }

    String buildUserAgent(Context context) {
        return String.format(
                "Piano composer SDK (Android %s (Build %s); %s %s/%s)",
                Build.VERSION.RELEASE,
                Build.ID,
                context.getResources().getBoolean(R.bool.isTablet) ? "Tablet" : "Mobile",
                Build.MANUFACTURER,
                Build.MODEL
        );
    }

    Composer getComposer() {
        return composer;
    }

    static void init(@NonNull Context context, @NonNull String aid, @Nullable String endpoint) {
        DependenciesProvider localInstance = instance;
        if (localInstance == null) {
            synchronized (DependenciesProvider.class) {
                localInstance = instance;
                if (localInstance == null)
                    instance = new DependenciesProvider(context, aid, endpoint);
            }
        }
    }

    static DependenciesProvider getInstance() {
        if (instance == null) {
            throw new IllegalStateException("Piano Composer SDK is not initialized! Make sure that you initialize it");
        }
        return instance;
    }
}

File: composer/src/main/java/io/piano/android/composer/DependenciesProvider.kt
package io.piano.android.composer

import android.content.Context
import android.os.Build
import androidx.annotation.RestrictTo
import com.squareup.moshi.Moshi
import io.piano.android.composer.model.PcusContainer
import io.piano.android.composer.model.PprvContainer
import io.piano.android.consents.ConsentJsonAdapterFactory
import io.piano.android.consents.PianoConsents
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.create
import java.util.concurrent.TimeUnit

@RestrictTo(RestrictTo.Scope.LIBRARY)
internal class DependenciesProvider private constructor(
    context: Context,
    aid: String,
    endpoint: Composer.Endpoint,
    pianoConsents: PianoConsents?,
) {
    private val prefsStorage = PrefsStorage(context)
    private val userAgent = "Piano composer SDK ${BuildConfig.SDK_VERSION} (Android ${Build.VERSION.RELEASE} " +
        "(Build ${Build.ID}); ${context.deviceType()} ${Build.MANUFACTURER}/${Build.MODEL})"

    private val okHttpClient = OkHttpClient.Builder()
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .addInterceptor(UserAgentInterceptor(userAgent))
        .addInterceptor(AidInterceptor(aid))
        .addInterceptor(RequestPolicyInterceptor(prefsStorage))
        .addInterceptor(
            HttpLoggingInterceptor().setLevel(
                if (BuildConfig.DEBUG || isLogHttpSet()) {
                    HttpLoggingInterceptor.Level.BODY
                } else {
                    HttpLoggingInterceptor.Level.NONE
                }
            )
        )
        .build()
    private val moshi = Moshi.Builder()
        .add(ComposerJsonAdapterFactory())
        .add(EventJsonAdapterFactory())
        .add(CustomValuesJsonAdapter)
        .add(UnixTimeDateAdapter)
        .add(ConsentJsonAdapterFactory)
        .build()

    private val moshiConverterFactory = MoshiConverterFactory.create(moshi)

    private val composerApi: ComposerApi = Retrofit.Builder()
        .baseUrl(endpoint.composerHost)
        .client(okHttpClient)
        .addConverterFactory(moshiConverterFactory)
        .build()
        .create()

    private val generalApi: GeneralApi = Retrofit.Builder()
        .baseUrl(endpoint.apiHost)
        .client(okHttpClient)
        .addConverterFactory(moshiConverterFactory)
        .build()
        .create()

    private val edgeCookiesProvider = EdgeCookiesProvider(
        prefsStorage,
        pianoConsents,
        moshi.adapter(PprvContainer::class.java),
        moshi.adapter(PcusContainer::class.java)
    )

    internal val composer: Composer = Composer(
        composerApi,
        generalApi,
        HttpHelper(ExperienceIdsProvider(prefsStorage, PageViewIdProvider), prefsStorage, moshi, userAgent),
        prefsStorage,
        aid,
        endpoint,
        edgeCookiesProvider,
        pianoConsents
    )

    @Suppress("NOTHING_TO_INLINE")
    private inline fun Context.deviceType() = if (resources.getBoolean(R.bool.isTablet)) "Tablet" else "Mobile"

    companion object {
        @JvmStatic
        @Volatile
        private var instance: DependenciesProvider? = null

        @JvmStatic
        internal fun init(context: Context, aid: String, endpoint: Composer.Endpoint, pianoConsents: PianoConsents?) {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) {
                        instance = DependenciesProvider(context, aid, endpoint, pianoConsents)
                    }
                }
            }
        }

        @JvmStatic
        fun getInstance(): DependenciesProvider = checkNotNull(instance) {
            "Piano Composer SDK is not initialized! Make sure that you initialize it"
        }
    }
}
